### ISSUE TITLE
Flag starred expressions in `return` and `yield`

### ISSUE THREAD

**Issue by ntBre** [astral-sh/ruff#16520]
Both of these are invalid on every Python version I've tried and should be flagged as `ParseError`s in the parser:

```python
def f(): yield *rest
def f(): return *rest
```

They result in `SyntaxError: can't use starred expression here` on 3.8 and 3.13 and the less helpful `SyntaxError: invalid syntax` on 3.7.

This is separate from the related case in #16485 where these are part of a tuple, which is allowed after 3.8.

_Originally posted by @ntBre in https://github.com/astral-sh/ruff/issues/16485#issuecomment-2701182623_

**Comment by AlexWaygood** [comment]
Python's `ast` module parses these:

```pycon
Python 3.11.4 (main, Sep 30 2023, 10:54:38) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ast
>>> ast.parse('def f(): return *x')
<ast.Module object at 0x7d5c2433d0>
```

which indicates to me that this syntax error is detected by CPython's compiler rather than CPython's parser. That means that this issue is a sub-issue of https://github.com/astral-sh/ruff/issues/11934. 

It doesn't mean that we shouldn't detect the syntax error (we absolutely should!). But it does mean that we may want a way of suppressing the diagnostic internally so that we can continue to test our parser directly against CPython's parser using e.g. the `py-fuzzer` script (which uses Python's `ast` module to generate random code).

**Comment by MichaReiser** [comment]
Thanks @AlexWaygood That would suggest that this should be implemented as a semantic syntax error, outside the parser.

**Comment by ntBre** [comment]
I'll start moving some of the other errors from #6591 to #11934 too. I think the `global` declarations in `try` nodes are another example:

```pycon
Python 3.12.9 (main, Feb 12 2025, 14:50:50) [Clang 19.1.6 ] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ast
>>> ast.parse("""
... a=5
...
... def f():
...     try:
...         pass
...     except:
...         global a
...     else:
...         print(a)
... """)
<ast.Module object at 0x79c0c0e06d10>

**Comment by AlexWaygood** [comment]
Good call. Some of the existing errors listed in https://github.com/astral-sh/ruff/issues/11934 and its sub-issues may also have been introduced in recent versions of CPython; we'll need to check when implementing those semantic syntax errors whether they should be emitted for all target versions or not

**Comment by dhruvmanila** [comment]
(I think from Python 3.9, you can test the CPython parser directly using `python -m ast path/to/file.py`.)

**Comment by ntBre** [comment]
This issue may also include `for` statements as added in 3.9 ([BPO](https://github.com/python/cpython/issues/90881)).

``` pycon
Python 3.13.1 (main, Dec  4 2024, 18:05:56) [GCC 14.2.1 20240910] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ast
>>> ast.parse("for _ in *rest:...")
<ast.Module object at 0x7cd78d9bb450>
>>> for _ in *rest: ...
  File "<python-input-2>", line 1
    for _ in *rest: ...
             ^^^^^
SyntaxError: can't use starred expression here
>>>
```
